### ISSUE TITLE
ELN Causing Lag

### ISSUE THREAD

**Issue by codgamerMC** [Electrical-Age/ElectricalAge#857]
So I opened this problem about 8 months ago and it still is unsolved, so let me go to the point, I have been using ELN for 11 months now, 6 months ago I really got rolling with the problems and expanding this simulation beyond what anyone can imagine, so I got a pretty big map about 1.46 GB in size 7500 blocks by 2500 blocks I would say is the size of the grid. The Grid is composed of miles of transmission and distribution lines, as many as about 8 Transmission stations and 10 substations, really big, this grid is composed of 8 Power Plants which make the system capacity of 1 MW or 1000 kW, the average consumption of the grid is 400 kW, really big I tell you, but since the system capacity was 100 kW, 10 times less than what it is now, I started getting TPS Problems and now they are flat bad, around 4-6, I removed the mod and TPS went to 20 so I know it's the mod what is causing the lag, how can I optimize it to make better use of it? I really love the mod and would love for it to be lag free, how could I use another CPU Core to optimize it, I know MC runs on Single Threads but are their anyways to set ELN on one side for the processor to process all the classes and codes it has to run? Any help I would appreciate!! Below some pictures of this monstrosity of grid and TPS.
![2018-06-02_09 08 10](https://user-images.githubusercontent.com/10374236/41064053-6cab04ce-69a8-11e8-96be-1850bf4c2c08.png)
![2018-06-01_13 14 29](https://user-images.githubusercontent.com/10374236/41064065-7722c086-69a8-11e8-89ee-57ff282749b1.png)
![2018-05-31_19 21 49](https://user-images.githubusercontent.com/10374236/41064071-7db2d8aa-69a8-11e8-8238-18c4e41ba492.png)
![2018-05-30_10 59 51](https://user-images.githubusercontent.com/10374236/41064131-9e98edf2-69a8-11e8-884e-c3a8c72da050.png)
![2018-05-28_19 39 09](https://user-images.githubusercontent.com/10374236/41064175-bfb08d92-69a8-11e8-906d-74fb464b2535.png)

**Comment by Baughn** [comment]
There really isn't a solution for this, beyond rewriting it to use multithreading and/or GPU acceleration.

You're flat-out exceeding the design limitations.

**Comment by codgamerMC** [comment]
I suppose that rewriting for it to accept multi threading is a complicated process right? That will take months or maybe year? I got two other questions, isn't their a way I can dedicate a core from my processor to run the mod? Is their a way to make signal repeaters more efficiently? Because I need signals to travel like 5000 blocks across the map and they tend to cause most of the lag in the TPS for what I have seen, thanks for the help @Baughn I really appreciate it because this is my favorite mod of all minecraft and what I consider to also be the best!

**Comment by jrddunbr** [comment]
Depending upon your operating system, you can pin an entire process to a single processor core. You can't just pin the Electrical Age mod to a process, since it's part of the Java run-time process.

That said, if you have a processor that can turbo a single core, this can be beneficial to performance.

On a side note:

I also wrote some code a while back that allows you to use OpenComputers (probably not ComputerCraft) linked cards, which don't require chunks to be loaded between the computers (it can even work from the overworld to the nether for example), you only need the chunks with the computers in it to be loaded.

https://github.com/jrddunbr/oc-eln/tree/master/elnProbeLink

To use it, all you need to do is craft a linked card pair, place one card in each computer, and then attach an Electrical Age Computer Probe to each computer. It will transfer the signals from one computer to the other. If you need help, I can create better instructions.

**Comment by codgamerMC** [comment]
Thanks man, some info is my OS is Windows 7 Ultimate 64 bit and processor is i7-4790k which has beats single core performance, so how can I get more out of it like you said, about the signals I will try it out

**Comment by codgamerMC** [comment]
Also, I would like to know if there is a software for OC to make a
monitoring station that you can acess the date from firefox or chrome? That
would be cool if there were such software, thanks for the help man!

**Comment by codgamerMC** [comment]
Also when I try executing the server or client lua errors just pop up

**Comment by jrddunbr** [comment]
Alright, so I looked into stuff a bit. The 4790K does have overclocking on a "per core" basis, but unfortunately, from what I've read, it only supports overclocking on random cores at a time. I don't have any experience with that, so you're on your own there with pinning a clock to a core.

If you can figure it out, Google will help you more than I can, but the (somehwat ineffective) method I used to use when I ran Windows XP was to use the process affinity settings in Task Manager. It would probably involve setting the Java process to a core, and the rest of your (major) processes to other cores. One piece of advice I do have here is that the system utilization of core 0 is always the most (due to the way Windows is designed) - which will make that core less favorable for such a setup - I'd select a core on cores 2-4.

There may be a better way to do this, and generally, doing complex things like that go a bit better on custom Linux boxes. I can't say that you'd see any amazing performance jump in any way there, so I wouldn't leap out of your way to do anything here, as it's purely speculative, and may have no effect, or negative effect.

That aside, "access the date from firefox or chrome"? Did you mean an application to monitor how the Electrical Age stuff is performing from a web browser?

I have been working on such a program. It's not done, but I've made some good progress, that's also on my GitHub. I could work on it a bit tonight, since I just finished a very related project two nights ago, where I was showing information in a browser from a different computer, and have worked out many of the kinks of making applications for OpenComputers.

https://github.com/jrddunbr/eln-oc-flask

What lua errors? Logs? (put them in tripple backticks or on https://pastebin.com

**Comment by codgamerMC** [comment]
First of all the program I was referring to by monitoring ELN stuff from a browser that's what I meant with that like design some monitoring programs that let's say I can monitor power plant kw production and that, that would be cool to see and second of all can you see the picture I sent or do I have to send it thru paste bin?

**Comment by jrddunbr** [comment]
Besides the first issue post, I see no images. If you attached them in an email, they got lost.

**Comment by codgamerMC** [comment]
Here is the Pastebin data plus a picture
https://pastebin.com/mNcWDEEF

**Comment by codgamerMC** [comment]
I have a question and is something that relatively pisses me off, why do transformers from the poles and the grid transformers have a huge capacitance in them, to raise voltage is really hard and when connecting lines to grid transformer they lost certainly cause blackouts if not burnout of transmission lines, is there a way to eliminate this and make it more realistic?

**Comment by jrddunbr** [comment]
I saw the pastebin, but not the image. Upload it to an image service like imgur, or view the GitHub issue and upload the image there.

The error is pretty cryptic - there is no > or < on that line, the line before, or the line after. I would check to make sure that you copied it right - you can use wget with an internet card. Just run the following:

Client:
```
wget https://raw.githubusercontent.com/jrddunbr/oc-eln/master/elnProbeLink/client.lua
```

Server:
```
wget https://raw.githubusercontent.com/jrddunbr/oc-eln/master/elnProbeLink/server.lua
```

That will download it to the computer. 

As for why transformers of larger size using lots of power like that, there's a simple explanation. As an optimization step, Baughn added buffers to the transformers over a certain size. What this does is add a buffer that scales to the size of the voltage - and this is really easy to see with T2 poles. They are still very much so a development feature, and not part of the base mod, so there's not much we're going to do to support it. Effectively, it makes the transformers act like large capacitors.

The reason we do this is very technical and related to the simulation. As you add more components to a grid, the linear equation that is used to solve the MNA grows in time complexity at a rate that is more than linear. By using "isolated" transformers over a certain size, this allows the computations to be split into multiple parallel computations, which complete in linear time, since the more things you add, the more calculations need to be made. If we didn't do this, then as you make your grid bigger and bigger, the linear equation required becomes much and much larger, and then Eln can no longer run at a TPS of 20, causing the game to lag a lot. Adding a bit of capacitance allows there to be some slack between the abbutment of the two equations - it doesn't have to be perfect, since calculations like that are sometimes not perfect.

I could go in more depth, but this basically is what is going on. **TL;DR** If we didn't add capacitance to the larger transformers, Eln would kill the TPS in no-time. There is currently no proposed fix; the obvious solution is to make the buffer smaller, but since Baughn is updating the code to 1.12, we won't see that until a newer version is released.

The best way to handle it is to use current limiting electronics - place an inline electrical probe in line with a high voltage relay and a inductor at full inductance. Set the probe to turn off the relay at higher than 2A. Once the voltage is the same on both sides, you can use a set of relays to carry the load, so that it's not over one cable.

**Comment by codgamerMC** [comment]
First of all happy Father's Day!! Second I did use the wget but that happened I will try to post the image as soon as I can, as for the capacitance, it kind of gives me headaches when connecting many poles with transformers together or wanting to raise the voltage from 0 to 13.2 in an instant it makes things blowup and when adding grid transformers I essentially have to use creative source to create the voltage of the system so I can safely put them together lot of work, I would like to see eln be more efficient as currently I am so in love with it and simulating getting bigger and bigger and TPS only getting worse and worse, I really got in love with electrical engineering at this point, final question, when do you think the 1.12 version will come out, do you think it will solve most of my problems?

**Comment by codgamerMC** [comment]
Link to the picture: https://imgur.com/KibXd7L

**Comment by jrddunbr** [comment]
Happy fathers day to you as well!

Huh. That's odd. I'll have to test the code again. The poles are doing that because you are putting too many amps down the wire, and they melt. You need to have surge protection circuits and other hardware in place - It's not as simple as just ploping down some poles and going "Oh, this will never explode!" Even I sometimes have explosions.

1.12 is a while out, but we wil be fixing as many bugs as possible. I would say most likely not before the end of summer in the earliest. I would say that by spring, we *should* have some kind of development builds running on 1.12. No gurantees

**Comment by codgamerMC** [comment]
I know things explode, I have like a dozen of breakers in every entry and
out point but you can't place breakers in transmission lines which means
that they can explode completely, If there was some breaker between line
and transformer that wouldn't happen, idk if you understand what I mean
but, simple breakers for 16 kW won't really help as those melt too, but
since the system has a capacity of 1056 kW so really there is no way to
protect the lines other than at their entry points but yet again lot of
Power Plants so that won't really help other than a breaker between
transformer and line like in line fuses in real life but I guess that is
complicated to do other than it being a place-able item between line and
transformer, I would like to see the bug fixes and other items added, this
is a really cool mod, happy fathers day again!!

**Comment by jrddunbr** [comment]
Generally, the way I do it is step down the power to 3.2kV and do the stuff
I need. Obviously, it would be great if we had breakers on poles, but we're
not that advanced yet.

If you need to have separate T2 feeds, you're using a lot of power, enough
so that you will probably encounter the sound manager bug. That's worse.

If you're creating   >150kW, you should really have a monitoring system
that works quickly over long distances to calculate overloads. It's pretty
hard to make that much power.

(Note to self: finish and document high power software for OC)

sent from Jared's phone

**Comment by codgamerMC** [comment]
Yeah few lines carry about 90 kW from south to north but I made a high amp custom breaker using relays which work great the system can generate 1056 kW and uses about 445.42 kW so lots of power going around the grid, so requieres a lot of stepping down to get relays in between lines, I would might send some pics so you can see that, another thing would it be to much to ask to make ELN Alternating Current and not Direct Current? Would that make the game unplayable like lagging due to calculations?

**Comment by jrddunbr** [comment]
The short answer about AC is that modern computers need to get more
powerful...

**Comment by codgamerMC** [comment]
Yeah I figured, since AC would require a lot of buffering and calculation,
although it would be extremely cool to see AC cause that would just make
the game even more complicated which is cool and having 3 phase systems and
what not, well only if there were stronger computers, thanks man for all
the help, when you revised the OC ELN program let me know so I can download
it and try it out.

**Comment by Baughn** [comment]
It would be possible to implement AC without taking that much more compute power. We'd need to track phase in addition to voltage, but there's no need to compute AC as if it is DC -- or rather, it can be treated mostly like DC, without needing to compute the constantly varying voltage.

But we won't, because there isn't very much benefit, and it would be a *little* more expensive. If the grid were rebuilt from scratch, today, there's a good chance we'd go with HVDC instead of AC.

As for the T2 transformer capacitance, the overly high capacitance is in fact why it hasn't been merged yet. If you use pre-release, unmerged code, the consequences are on you... I'd welcome patches to fix it, if you can figure out how, but I'm not spending time on that right now.

**Comment by codgamerMC** [comment]
Phase AC would be cool to see, but that's the developers choice on what's more efficient in terms of the load it causes on the server as for the capacitance, I noticed that since I started using the T2 dev release but the high capacitance hasn't destroyed my big ass grid, that's because I am always careful with what I do in terms of them converters but I managed to work it out, one time one grid converter destroyed part of a transmission line but because having breakers in each end it didn't destroy other parts of the line, but it's workable but just some suggestions, if the capacitance could be removed without it being less efficient that would be cool and also an inline variable breaker amperage depending on voltage would be cool (meaning a breaker between line and transformer like in real life substations, would make it more efficient in my opinion because you don't have to down it all until it's 3.2 then put regular breakers sometimes up to 6-10 per line and then upping it all again) lots of calculations poles transformers and also voltage loss due to resistance because of the many breakers and stuff, thanks for everything.

**Comment by jrddunbr** [comment]
Hmm, I should look at the T2 code if that's all that's keeping it from being merged... Sounds like something I *may* be able to handle. I'll try to fix a few bugs if I get a chance as well...

EDIT: Oh boy, right into the MNA we go... This can't end well.

**Comment by jrddunbr** [comment]
I can't replicate the bug you were having with OpenComputers. I'm not even sure how that bug is coming about, since there is no `<` character in the `client.lua` file...

Make sure you're using the latest version of OpenComputers for 1.7.10.

**Comment by codgamerMC** [comment]
I am using the latest version, just upgraded it and still encountering the
same problem

**Comment by codgamerMC** [comment]
On a side note Jared, see this picture (https://imgur.com/5FUywwl) this is
what I meant about a monitoring system that you can see from a browser,
like you can customize it and draw a line to monitor the system or a
certain line, being linked to lets say one of the sides of the ELN Probe so
it displays the information, that would be cool, so I don't have to use all
those signal transmitters for a station in minecraft, since it might be
done from the web, I hope it wouldn't be to difficult to do, I suspect it
is, let me know, thanks for all the help Jared!!

**Comment by jrddunbr** [comment]
Whoa, that's a bit more advanced than what I was thinking.

I'll see what I can do. Seems like a cool thing to do.

**Comment by codgamerMC** [comment]
Hey I got it working!! It was the outdated OpenOS built in from the mod so
I installed it else where and it worked so I am going to do some testing
and let you know, thanks for everything, I know the picture I sent you is
quite advanced but its cool, let me know what else you think you can do if
not that's fine, its not something I really need, but more like a cool
addition

**Comment by codgamerMC** [comment]
Hey Jared, I find a big problem with the program from OC, I find that
everytime I turn off the server and when I turn it back on, the program has
crashed in all computers saying things like probe not found and I have to
restart over 70 computers to get the monitoring system to work correcctly,
from server side computers to client side, any ideas as to why?

**Comment by jrddunbr** [comment]
I've had this problem on my server as well. The documentation for OpenComputers says that you need to call sleep(0) or some other interrupt at least once every 5 seconds. I've done this to no avail.. I'll have to look into it more in the coming days

**Comment by Baughn** [comment]
> I've had this problem on my server as well. The documentation for OpenComputers says that you need to call sleep(0) or some other interrupt at least once every 5 seconds. I've done this to no avail.. I'll have to look into it more in the coming days

The problem is, during a restart it's likely that a single tick (the first one) will take up to multiple minutes.

This triggers OpenComputers' lag protection, and crashes the program. There's no fix that I know of.

**Comment by jrddunbr** [comment]
Hmm, thanks for the insight. I'll try to get around that... somehow
